### DISCUSSION TITLE
Install Redash version 10.0.0, without Docker, error!

### DISCUSSION THREAD

**Discussion by SlavchoSpasenoski** [getredash/redash#5955] [Q&A] [answered]
Hello,

I am trying to install **Redash version 10.0.0 without Docker**, on Ubuntu 20.04. 
For some security reasons we have removed docker from our environments, so We need to rewrite and reconfigure Redash script with version 10.0.0.

There is a lot dependencies, (also we are forced to use python 3.8 or above) so I created custom Requirements.txt and Requirements_all_ds.txt files, with some upgrades, some of these is excluded (mostly data sources) and so on.. but they all executable so far.

**What I did until now:** I trying to modify the old script (until Redash v5) that is worked without Docker
**Here is a link with script from forum discussion**  -> https://discuss.redash.io/t/bring-back-bootstrap-sh/2436/12

**Problems that I faced so far:**

After execution script, I executing commands to run server and workers, instead old supervisord.conf script with celery, I tried to execute following command with **"rq"** s:

Web server: ./manage.py runserver --debugger --reload
RQ: ./manage.py rq worker
RQ Scheduler: ./manage.py rq scheduler

( followed this guide also: https://redash.io/help/open-source/dev-guide/setup )

But when I writte localhost:5000 on my web browser I see, this:

![image](https://user-images.githubusercontent.com/28984924/228822774-756e0337-4fae-4e12-9c3a-25c0e67cd505.png)

Also, when I write localhost:6379, for redis, I got this error

![image](https://user-images.githubusercontent.com/28984924/228822925-6bf75196-bfd8-4f81-8fa9-1498e8ef36ed.png)

After click, **setup**, I've got "Internal Server Error" on my browser. This is Happened on forum discussion (link from above) but @arikfr  gave a suggestion to use [version.redash.io/api/releases](http://version.redash.io/api/releases) , not repo tarball, but there is not for v10.0.0 ?!

Main question is, is it possible at all to install redash 10 without docker on Ubuntu 20.04, if it true, do Redash going to be stable, manageable, and no have security flaws?!

Here is mine script:
[install_redash10.0.txt](https://github.com/getredash/redash/files/11110854/install_redash10.0.txt)

Here is mine custom Requirement files

[requirements_all_ds.txt](https://github.com/getredash/redash/files/11110857/requirements_all_ds.txt)
[requirements.txt](https://github.com/getredash/redash/files/11110858/requirements.txt)

Any suggestion will appreciate,

Thanks, Slavcho

**Comment by justinclift** [comment]
> I've got "Internal Server Error" on my browser.

When that happens, the details for the error (eg Python error output) should be shown in one of the terminal windows that you are running.

You'll need to go through the cycle of finding that error, working out what that means + fix it, then try again.

It'll probably take you a few days to make it all work nicely, so don't give up yet. :smile:

**Reply by SlavchoSpasenoski** [reply]
Thanks for answer, I've successfully run redash server without docker, I have some problems with Yarn to run server & workers, but I resolved with using master repo, not version 10.0.0 release.

So now, I facing another error, with data source connection and this is:

![image](https://user-images.githubusercontent.com/28984924/229497854-7be7ccc6-f0c6-4f60-a102-800f297c798b.png)

Here also log from workers:
![Screenshot_1](https://user-images.githubusercontent.com/28984924/229498028-a0c00937-798e-494c-b359-0c6bdfe06f22.png)

Thanks

**Reply by justinclift** [reply]
Yeah, using the `master` branch is the right starting point.

That error message:

```
  Key (ord_id)=(1) is not present in table "organizations".
```

... sounds like your database hasn't been initialised yet, and/or you haven't then gone through the very first "setup new user" process via the sign in page.

That's a wild guess though. :grin: 

If that's right, and you haven't created the initial tables yet, try this:

```
$ python3 manage.py database create_tables
```

That should initialise the database with the needed tables, which then lets the "first user setup" process run in the web interface.

Btw, are you using a Python virtual environment for this?

If not, make a note to look into that later as they're super useful.  It lets you try out Python modules in an isolated "environment" so they don't interfere with the rest of your system. eg:

```
$ python3 -m venv myenv1
$ source myenv1/bin/activate
(myenv1) $ pip3 install -r requirements.txt  (etc)
```

Anyway, hope some of the above is useful. :smile:

**Reply by SlavchoSpasenoski** [reply]
Logically it shouldn't throw this kind of error if tables are not created, because error is constraint, but magically I just re-run create_tables script and wroks fine, now I am able to connect with local instance of MySQL as data source, but I facing another error

And this is following:

![image](https://user-images.githubusercontent.com/28984924/230133200-bdb75d50-8864-4cbe-9fb0-54a6b6fe6f4a.png)

Just to note, env file is properly filled and configured in correct path, so now I temporary hardcoded REDASH_COOKIE_SECRET into /redash/settings/ _ _init_ _.py file to be able to run redash server & workers.

Mine script is following:




[Redash_install_withoutDocker.txt](https://github.com/getredash/redash/files/11160614/Redash_install_withoutDocker.txt)

Note: I use custom requirements.txt and requirements_all_ds.txt files to satisfy Python 3.8 and Ubuntu 20.04 dependencies.

**Reply by justinclift** [reply]
Ahhh, cool.  You're making solid progress. :smile:

Btw, while hard coding that environment value will work, it's not a fantastic approach. :wink:

From lines 66 to 70 of that same file:

https://github.com/getredash/redash/blob/ad7d30f91de64a291eb89892c713bd0067c47ecc/redash/settings/__init__.py#L66-L70

What those lines are saying is that Redash will look for `REDASH_COOKIE_SECRET` in the running program environment, and use whatever value is there.  And if it's missing, it'll error out with the message you saw.

That `.env` file though is only used for Docker Compose:

https://github.com/getredash/redash/blob/ad7d30f91de64a291eb89892c713bd0067c47ecc/docker-compose.yml#L11-L12

So when running without Docker, you'll need to set the environment variables some other way before starting Redash.

Commonly, that's done by adding them to your shell (eg `bash`) login script.  With a default Ubuntu 20.04 install, that's er... hmmm.  It'll either be `.bashrc` or `.bash_profile`.  I'm not remembering off the top of my head. :wink:

Either way (or just do both to make super sure), you'd include it like:

```
export REDASH_COOKIE_SECRET=some_value_here
```

Then re-start your bash session.  eg open a new Terminal windows.  Starting Redash after that should find the environment variable, and work fine.  The other environment variables are set the same way.

**Comment by justinclift** [comment]
Oh, actually looking at your `Redash_install_withoutDocker.txt`, that's a bash script.  You already have a section at the top which sets environment variables, so you could just add the cookie secret there.

On a side note, lines 86-103 already do some environment variable setup.  Instead of having them add to the `$REDASH_BASE_PATH/env`, you could have them add to the `.bashrc` or `.bash_profile` instead.

Next thought... it seems to be relying on an ancient version of Redash files.  Version 4.0, in line 11.  Probably better to bump that version number up a bit. :smile:

Hmmm, but line 119 is saying Redash 10.0, so maybe those files do work well enough anyway. :smile:

With the Node version 14 on line 121, I'm pretty sure Node 16 is the better choice when using the `master` branch.

Skimming over the rest of it, those are the only things jumping out at me.  Hopefully some of that helped. :smile:

Btw, please don't read the above as being too critical.  I reckon you're doing good stuff, and it'll be helpful for others. :smile:

**Comment by SlavchoSpasenoski** [comment, marked as answer]
Great suggestion, .bashrc is the correct script with global variables, but solution is very simple, more is my fault, than error... So

While executing command "manage.py database create-tables" need be runed with bin/run ...

The correct call is: sudo -u redash **bin/run** ./manage.py create-tables

Until now everything works fine and redash without docker is up and running.

Thank you for your effort @justinclift

**Reply by justinclift** [reply]
You're welcome.  It's been one of those "someone should write this stuff down" things for ages. :smile:

Now that you have it working, are you going to put those pieces in some form of systemd unit file(s), so it all starts up automatically?

You know... just for the extra challenge. :grin:

**Reply by SlavchoSpasenoski** [reply]
Yeah, I'm on It... 

I will post here final script when it's Done.

Thanks again @justinclift

**Reply by icdw** [reply]
@SlavchoSpasenoski appreciate if you can post the complete scripts, etc., that you made for Redash to work without Docker

Thank you very much in advance!
